File: tf2_ros/test/test_buffer_client.cpp
/*
 * Copyright (c) 2019, Open Source Robotics Foundation, Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of the copyright holder nor the names of its
 *       contributors may be used to endorse or promote products derived from
 *       this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <chrono>
#include <future>
#include <thread>

#include <gtest/gtest.h>

#include <tf2_msgs/action/lookup_transform.hpp>
#include <rclcpp/rclcpp.hpp>
#include <rclcpp_action/rclcpp_action.hpp>

#include <tf2_ros/buffer.h>
#include <tf2_ros/buffer_client.h>

static const std::string ACTION_NAME = "test_tf2_buffer_action";

class MockBufferServer : public rclcpp::Node
{
  using LookupTransformAction = tf2_msgs::action::LookupTransform;
  using GoalHandle = rclcpp_action::ServerGoalHandle<LookupTransformAction>;

public:
  MockBufferServer()
    : rclcpp::Node("mock_buffer_server"),
      transform_available_(false)
  {
    action_server_ = rclcpp_action::create_server<LookupTransformAction>(
      get_node_base_interface(),
      get_node_clock_interface(),
      get_node_logging_interface(),
      get_node_waitables_interface(),
      ACTION_NAME,
      [](const rclcpp_action::GoalUUID &, std::shared_ptr<const LookupTransformAction::Goal>)
        {return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;},
      [](const std::shared_ptr<GoalHandle>){return rclcpp_action::CancelResponse::ACCEPT;},
      std::bind(&MockBufferServer::acceptedCallback, this, std::placeholders::_1));
  }

  void acceptedCallback(const std::shared_ptr<GoalHandle> goal_handle)
  {
    // Simulate doing some work in here; otherwise, we can complete the goal
    // before the rclcpp_action interface ever has time to do any work.
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    auto result = std::make_shared<LookupTransformAction::Result>();
    if (transform_available_) {
      result->transform.transform = transform_;
      goal_handle->succeed(result);
    } else {
      goal_handle->abort(result);
    }
  }

  void setTransform(const geometry_msgs::msg::Transform & transform)
  {
    transform_ = transform;
    transform_available_ = true;
  }

private:
  geometry_msgs::msg::Transform transform_;
  bool transform_available_;

  rclcpp_action::Server<LookupTransformAction>::SharedPtr action_server_;
};

class TestBufferClient : public ::testing::Test
{
protected:
  static void SetUpTestCase()
  {
    rclcpp::init(0, nullptr);
  }

  static void TearDownTestCase()
  {
    rclcpp::shutdown();
  }

  void SetUp()
  {
    node_ = std::make_shared<rclcpp::Node>("tf_buffer");
    client_ = std::make_unique<tf2_ros::BufferClient>(node_, ACTION_NAME);
    mock_server_ = std::make_shared<MockBufferServer>();

    executor_.add_node(node_);
    executor_.add_node(mock_server_);

    // Start spinning in a thread
    spin_thread_ = std::thread(
      std::bind(&rclcpp::executors::SingleThreadedExecutor::spin, &executor_));

    // Wait for discovery
    ASSERT_TRUE(client_->waitForServer(std::chrono::seconds(10)));
    // This hack ensure the action server can see the client
    // TODO(jacobperron): Replace with discovery rclcpp API when it exists
    while (mock_server_->count_subscribers("/" + ACTION_NAME + "/_action/feedback") < 1u) {
      std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }
  }

  void TearDown()
  {
    executor_.cancel();
    spin_thread_.join();
    mock_server_.reset();
    client_.reset();
    node_.reset();
  }

  rclcpp::Node::SharedPtr node_;
  std::unique_ptr<tf2_ros::BufferClient> client_;
  std::shared_ptr<MockBufferServer> mock_server_;
  rclcpp::executors::SingleThreadedExecutor executor_;
  std::thread spin_thread_;
};

TEST_F(TestBufferClient, lookup_transform_available)
{
  // Set a transform to be available
  geometry_msgs::msg::Transform transform_in;
  transform_in.rotation.w = 0.5;
  transform_in.translation.x = 42.0;
  mock_server_->setTransform(transform_in);

  auto transform_out = client_->lookupTransform(
    "test_target_frame", "test_source_frame", tf2::get_now());

  EXPECT_EQ(transform_out.transform.rotation.w, transform_in.rotation.w);
  EXPECT_EQ(transform_out.transform.translation.x, transform_in.translation.x);
}

TEST_F(TestBufferClient, lookup_transform_unavailable)
{
  EXPECT_THROW(
    client_->lookupTransform("test_target_frame", "test_source_frame", tf2::get_now()),
    tf2_ros::GoalAbortedException);
}

TEST_F(TestBufferClient, can_transform_available)
{
  // Set a transform to be available
  geometry_msgs::msg::Transform transform_in;
  transform_in.rotation.w = 0.5;
  transform_in.translation.x = 42.0;
  mock_server_->setTransform(transform_in);

  EXPECT_TRUE(client_->canTransform("test_target_frame", "test_source_frame", tf2::get_now()));
}

TEST_F(TestBufferClient, can_transform_unavailable)
{
  EXPECT_FALSE(client_->canTransform("test_target_frame", "test_source_frame", tf2::get_now()));
}

int main(int argc, char **argv){
  testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

File: tf2_ros_py/test/test_buffer_client.py
# Copyright 2019 Open Source Robotics Foundation, Inc.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above
#    copyright notice, this list of conditions and the following
#    disclaimer in the documentation and/or other materials provided
#    with the distribution.
#  * Neither the name of the Willow Garage nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

import pytest
import threading
import time

import rclpy

from tf2_ros.buffer_client import BufferClient
from geometry_msgs.msg import TransformStamped
from tf2_msgs.action import LookupTransform
from tf2_py import BufferCore, LookupException
from rclpy.executors import SingleThreadedExecutor
from rclpy.time import Time
from tf2_msgs.msg import TF2Error


def build_transform(target_frame, source_frame, stamp):
    transform = TransformStamped()
    transform.header.frame_id = target_frame
    transform.header.stamp = stamp
    transform.child_frame_id = source_frame

    transform.transform.translation.x = 42.0
    transform.transform.translation.y = -3.14
    transform.transform.translation.z = 0.0
    transform.transform.rotation.w = 1.0
    transform.transform.rotation.x = 0.0
    transform.transform.rotation.y = 0.0
    transform.transform.rotation.z = 0.0

    return transform


class MockBufferServer():
    def __init__(self, node, buffer_core):
        self.action_server = rclpy.action.ActionServer(node, LookupTransform, 'lookup_transform', self.execute_callback)
        self.buffer_core = buffer_core

    def execute_callback(self, goal_handle):
        response = LookupTransform.Result()
        response.transform = TransformStamped()
        response.error = TF2Error()

        try:
            if not goal_handle.request.advanced:
                transform = self.buffer_core.lookup_transform_core(target_frame=goal_handle.request.target_frame,
                                                                   source_frame=goal_handle.request.source_frame,
                                                                   time=Time.from_msg(goal_handle.request.source_time))
            else:
                transform = self.buffer_core.lookup_transform_full_core(
                    target_frame=goal_handle.request.target_frame,
                    target_time=Time.from_msg(goal_handle.request.target_time),
                    source_frame=goal_handle.request.source_frame,
                    source_time=Time.from_msg(goal_handle.request.source_time),
                    fixed_frame=goal_handle.request.fixed_frame
                )
            response.transform = transform
            goal_handle.succeed()
        except LookupException as e:
            response.error.error = TF2Error.LOOKUP_ERROR
            goal_handle.abort()

        return response

    def destroy(self):
        self.action_server.destroy()


class TestBufferClient:
    @classmethod
    def setup_class(cls):
        cls.context = rclpy.context.Context()
        rclpy.init(context=cls.context)
        cls.executor = SingleThreadedExecutor(context=cls.context)
        cls.node = rclpy.create_node('TestBufferClient', context=cls.context)
        cls.executor.add_node(cls.node)

        buffer_core = BufferCore()
        transform = build_transform('foo', 'bar', rclpy.time.Time().to_msg())
        buffer_core.set_transform(transform, 'unittest')

        cls.mock_action_server = MockBufferServer(cls.node, buffer_core)

    @classmethod
    def teardown_class(cls):
        cls.mock_action_server.destroy()
        cls.node.destroy_node()
        rclpy.shutdown(context=cls.context)

    def setup_method(self, method):
        self.spinning = threading.Event()
        self.spin_thread = threading.Thread(target=self.spin)
        self.spin_thread.start()

    def teardown_method(self, method):
        self.spinning.set()
        self.spin_thread.join()

    def spin(self):
        while self.context.ok() and not self.spinning.is_set():
            self.executor.spin_once(timeout_sec=0.05)

    def test_lookup_transform_true(self):
        buffer_client = BufferClient(
            self.node, 'lookup_transform', check_frequency=10.0, timeout_padding=rclpy.duration.Duration(seconds=0.0))

        result = buffer_client.lookup_transform(
            'foo', 'bar', rclpy.time.Time(), rclpy.duration.Duration(seconds=5.0))

        tf = build_transform('foo', 'bar', rclpy.time.Time().to_msg())
        assert tf == result
        buffer_client.destroy()

    def test_lookup_transform_fail(self):
        buffer_client = BufferClient(
            self.node, 'lookup_transform', check_frequency=10.0, timeout_padding=rclpy.duration.Duration(seconds=0.0))

        with pytest.raises(LookupException) as excinfo:
            buffer_client.lookup_transform(
                'bar', 'baz', rclpy.time.Time(), rclpy.duration.Duration(seconds=5.0))

        assert LookupException == excinfo.type

        buffer_client.destroy()
